### PR TITLE
Standardize SSL settings

### PR DESCRIPTION
**Pull request by edmocosta**
### What this PR does?
**Added the following SSL settings**
`ssl_enabled`: Enable/disable SSL
`ssl_client_authentication`: Controls the server’s behavior in regard to requesting a certificate from client connections
   
**Reviewed and deprecated the following SSL settings to comply with Logstash's [naming convention](https://github.com/elastic/logstash/issues/14905)**
`ssl` in favor of `ssl_enabled`
`ssl_verify_mode` in favor of `ssl_client_authentication`

**Other changes**

- Using `ssl_verify_mode` should keep the current behavior as it is. Once upgraded to `ssl_client_authentication`, it will validate and reject configurations with `ssl_certificate_authorities` set and `ssl_client_authentication = > none`.  The current `ssl_verify_mode` behavior is to silently ignore the `none` value and use `force_peer/required`.

- Fixed `ssl_peer_metadata` when `ssl_enabled => false`: https://github.com/logstash-plugins/logstash-input-beats/pull/470#discussion_r1153374813

- Added tests for the new and existing `SslContextBuilder` methods.

The behavior standardization across plugins, such as the accepted certificate formats, default values, etc will be tackled in future PRs.

---
Closes https://github.com/elastic/logstash/issues/14925
Closes https://github.com/logstash-plugins/logstash-input-beats/issues/399

### CODE REVIEW

File: lib/logstash/inputs/beats.rb [hunk 1]
```diff
@@ -275,20 +259,39 @@ def target_codec_on_field?
...
+    certificate_authorities_configured?
+  end
+
+  def certificate_authorities_configured?
     @ssl_certificate_authorities && @ssl_certificate_authorities.size > 0
   end
 
   def client_authentication_metadata?
-    @ssl_peer_metadata && ssl_configured? && client_authentification?
+    @ssl_enabled && @ssl_peer_metadata && ssl_configured? && client_authentication_enabled?
```
Comment (edmocosta) [inline lib/logstash/inputs/beats.rb:278]: I have added the `@ssl_enabled` here to fix an existing issue. When `ssl => false`, ` ssl_peer_metadata => true`, and `ssl_certificate_authorities` is configured, it tries to [extract the metadata ](https://github.com/logstash-plugins/logstash-input-beats/blob/main/lib/logstash/inputs/beats/message_listener.rb#L135) from the netty context and crashes as SSL is actually disabled.

File: src/main/java/org/logstash/netty/SslContextBuilder.java [hunk 2]
```diff
@@ -29,25 +28,41 @@
...
+                case OPTIONAL:
+                    return ClientAuth.OPTIONAL;
+                case REQUIRED:
+                    return ClientAuth.REQUIRE;
+                default:
+                    throw new IllegalArgumentException(
+                            String.format("Unsupported SSL client authentication mode `%s`", this.name().toLowerCase())
+                    );
+            }
+        }
```
Comment (andsel) [inline src/main/java/org/logstash/netty/SslContextBuilder.java:48]: Maybe we could have a more explicit mapping without using the switch with:
```java
  NONE(ClientAuth.NONE),
  OPTIONAL(ClientAuth.OPTIONAL),
  REQUIRED(ClientAuth.REQUIRE);
  
  private final ClientAuth clientAuth;

  private SslClientVerifyMode(ClientAuth clientAuth) {
      this.clientAuth = clientAuth;
  }
   
  public ClientAuth toClientAuth() {
      return clientAuth;
  }
```

File: src/main/java/org/logstash/netty/SslContextBuilder.java [hunk 3]
```diff
@@ -129,58 +162,34 @@ public static String[] getDefaultCiphers(){
...
-    }
 
     public SslContext buildContext() throws Exception {
-        io.netty.handler.ssl.SslContextBuilder builder = io.netty.handler.ssl.SslContextBuilder.forServer(sslCertificateFile, sslKeyFile, passPhrase);
-
         if (logger.isDebugEnabled()) {
-            logger.debug("Available ciphers: " + SUPPORTED_CIPHERS);
-            logger.debug("Ciphers:  " + Arrays.toString(ciphers));
+            logger.debug(String.format("Available ciphers: %s", SUPPORTED_CIPHERS));
+            logger.debug(String.format("Ciphers: %s", Arrays.toString(ciphers)));
```
Comment (andsel) [inline src/main/java/org/logstash/netty/SslContextBuilder.java:177]: We can leverage the log4j templating and avoid the move to String format


```suggestion
            logger.debug("Available ciphers: {}", SUPPORTED_CIPHERS);
            logger.debug("Ciphers: {}", Arrays.toString(ciphers));
```

File: src/main/java/org/logstash/netty/SslContextBuilder.java [hunk 4]
```diff
@@ -129,58 +162,34 @@ public static String[] getDefaultCiphers(){
...
-            builder.trustManager(loadCertificateCollection(certificateAuthorities));
-            if(verifyMode == SslClientVerifyMode.FORCE_PEER) {
-                // Explicitly require a client certificate
-                builder.clientAuth(ClientAuth.REQUIRE);
-            } else if(verifyMode == SslClientVerifyMode.VERIFY_PEER) {
-                // If the client supply a client certificate we will verify it.
-                builder.clientAuth(ClientAuth.OPTIONAL);
+        if (isClientAuthenticationEnabled(verifyMode)) {
+            if (logger.isDebugEnabled()) {
+                logger.debug(String.format("Certificate Authorities: %s", Arrays.toString(certificateAuthorities)));
```
Comment (andsel) [inline src/main/java/org/logstash/netty/SslContextBuilder.java:187]:
```suggestion
                logger.debug("Certificate Authorities: {}", Arrays.toString(certificateAuthorities));
```

File: src/test/java/org/logstash/netty/SslContextBuilderTest.java [hunk 5]
```diff
@@ -1,23 +1,76 @@
...
+import io.netty.buffer.ByteBufAllocator;
+import io.netty.handler.ssl.ClientAuth;
+import io.netty.handler.ssl.SslContext;
 import org.hamcrest.core.Every;
+import org.junit.Assert;
+import org.junit.Rule;
 import org.junit.Test;
+import org.junit.rules.ExpectedException;
 
+import static org.hamcrest.Matchers.*;
```
Comment (andsel) [inline src/test/java/org/logstash/netty/SslContextBuilderTest.java:12]: Avoid `*` import switch to explicit list

Comment (andsel) [reply src/test/java/org/logstash/netty/SslContextBuilderTest.java:12]: same for the following import of such kind

File: src/test/java/org/logstash/netty/SslContextBuilderTest.java [hunk 6]
```diff
@@ -37,4 +89,141 @@ public void getDefaultCiphers() throws Exception
...
+        assertThat(sslContextBuilder.getVerifyMode(), is(SslClientVerifyMode.NONE));
+    }
+
+    private void assertSetClientAuthenticationThrowWithNoCerts(SslClientVerifyMode mode) {
+        final SslContextBuilder sslContextBuilder = createSslContextBuilder();
+        final String expectedMessage = "Certificate authorities are required to enable client authentication";
+
+        try {
+            sslContextBuilder.setClientAuthentication(mode, new String[0]);
+            Assert.fail("No exception was thrown");
```
Comment (andsel) [inline src/test/java/org/logstash/netty/SslContextBuilderTest.java:139]: All the other `Assert` methods are statically imported, so maybe we could use the same pattern and move to `fail()`

**Review by andsel — CHANGES_REQUESTED** [review]
Overall seems good, but left some comments for improvement.

File: lib/logstash/inputs/beats.rb [hunk 7]
```diff
@@ -275,20 +259,39 @@ def target_codec_on_field?
...
   end
 
   def client_authentication_required?
-    @ssl_verify_mode == "force_peer" 
+    @ssl_client_authentication && @ssl_client_authentication.downcase == "required"
```
[the remark this answers is not in the document]
Comment (edmocosta) [reply lib/logstash/inputs/beats.rb:282]: Hi @andsel, thanks for reviewing! I don't see much gain as well so I will keep it as it's.

**Review by andsel — APPROVED** [review]
LGTM

**Pull request merged**